vulnerabilities/data_source.py

```diff
@@ -0,0 +1,262 @@
+        PackageURL objects and strings. As a convention, the former is referred to in variable names, etc. as
+        "package_urls" and the latter as "purls".
+    """
+    summary: str
+    impacted_package_urls: Sequence[PackageURL]
+    resolved_package_urls: Sequence[PackageURL] = dataclasses.field(default_factory=list)
+    references: Sequence[str] = dataclasses.field(default_factory=list)
+    cve_id: Optional[str] = None
+
+    @property
+    def impacted_purls(self) -> Set[str]:
+        return {str(p) for p in self.impacted_package_urls}
+
+    @property
+    def resolved_purls(self) -> Set[str]:
+        return {str(p) for p in self.resolved_package_urls}
+
+
+@dataclasses.dataclass
+class DataSource(ContextManager):
+    """
+    This class defines how importers consume advisories from a data source.
+
+    It makes a distinction between newly added records since the last run and modified records. This allows the import
+    logic to pick appropriate database operations.
+    """
+    batch_size: int
+    cutoff_date: Optional[datetime] = None
+    config: Optional[Mapping[str, Any]] = dataclasses.field(default_factory=dict)
+
+    def __enter__(self):
+        """
+        Subclasses acquire per-run resources, such as network connections, file downloads, etc. here.
+        """
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """
+        Subclasses release per-run resources acquired in __enter__() here.
+        """
+        pass
+
+    def validate_config(self) -> None:
+        """
+        Subclasses can check self.config for mandatory parameters here.
+
+        The method is called in Importer.make_data_source(). In case of errors, subclasses should raise
+        InvalidConfigError.
+        """
+        pass
+
+    def added_advisories(self) -> List[Advisory]:
+        """
+        Subclasses yield batch_size sized batches of Advisory objects that have been added to the data source
+        since self.cutoff_date.
+        """
+        raise StopIteration
+
+    def updated_advisories(self) -> List[Advisory]:
+        """
+        Subclasses yield batch_size sized batches of Advisory objects that have been modified since
+        self.cutoff_date.
+
+        NOTE: Data sources that do not enable detection of changes to existing records vs added records must only
+              implement this method, not new_records(). The ImportRunner relies on this contract to decide between
+              insert and update operations.
+        """
+        raise StopIteration
+
+    def error(self, msg: str) -> None:
+        """
+        Helper method for raising InvalidConfigError with the class name in the message.
+        """
+        raise InvalidConfigError(f'{type(self).__name__}: {msg}')
+
+
+class GitDataSource(DataSource):
+    repository_url: Optional[str] = None
+    branch: str = 'master'
```

Comment:
This will probably give an error if the user doesn't specify the branch to checkout and the repo doesn't have a `master` branch. 

I think `branch` should not have any default value.  The pygit2's docs at https://www.pygit2.org/repository.html#pygit2.clone_repository mentions  regarding `checkout_branch`  
`Branch to checkout after the clone. The default is to use the remote’s default branch.`

Reply:
Thanks for pointing that out! You are right, the correct behaviour is to use the repository's default branch. I just realized that today as well and already changed it.